llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp

```diff
@@ -4426,6 +4511,56 @@ SDValue DAGTypeLegalizer::WidenVecRes_Binary(SDNode *N) {
                      {InOp1, InOp2, Mask, N->getOperand(3)}, N->getFlags());
 }
 
+SDValue DAGTypeLegalizer::WidenVecRes_CMP(SDNode *N) {
+  LLVMContext &Ctxt = *DAG.getContext();
+  SDLoc dl(N);
+
+  SDValue LHS = N->getOperand(0);
+  SDValue RHS = N->getOperand(1);
+  EVT OpVT = LHS.getValueType();
+  EVT TransformedOpVT = TLI.getTypeToTransformTo(Ctxt, OpVT);
+  if (TransformedOpVT.isVector() &&
+      TransformedOpVT.getVectorNumElements() > OpVT.getVectorNumElements()) {
+    LHS = GetWidenedVector(LHS);
+    RHS = GetWidenedVector(RHS);
+  }
+
+  EVT WidenResVT = TLI.getTypeToTransformTo(Ctxt, N->getValueType(0));
+  ElementCount WidenResEC = WidenResVT.getVectorElementCount();
+  EVT WidenResElementVT = WidenResVT.getVectorElementType();
+
+  // At this point we know that the type of operands should be legal,
```

Comment:
This doesn't sound right. The result types are legalized first, so we don't know at this point that the operands are legal.

Reply:
Perhaps I've worded this comment a bit poorly. What I meant is that since (unless I'm massively misunderstanding something) the node that produces a value is legalized strictly before any nodes that use this value, the actual `SDValue`s of the operands that we retrieve a few lines above this comment are guaranteed to be correctly widened. I am not 100% confident that those types would be fully legal (i.e. the type wouldn't be something like <4 x i29>), but it shouldn't matter in this scenario I believe since `WidenVecRes_CMP` makes sure that the CMP node is retyped so that the return type and operand types match, so even if any further legalization is required after this point, the fact that the result and operand types are the same prevents us from running into legalization cycles. Hope this clarifies things a bit, and I will rewrite that comment a bit better shortly.

Reply:
> Perhaps I've worded this comment a bit poorly. What I meant is that since (unless I'm massively misunderstanding something) the node that produces a value is legalized strictly before any nodes that use this value, the actual SDValues of the operands that we retrieve a few lines above this comment are guaranteed to be correctly widened.

The legalization happens from the root of the SDAG to the leaves. This means that nodes using a value are legalized *before* nodes producing the value.

I think your general point still stands because you explicitly look up the type the operands will get widened to.

Reply:
> This means that nodes using a value are legalized before nodes producing the value.

I don't think I agree with this -- as far as I'm aware it's the direct opposite, a node can only be legalized after all of its operands are legalized. That's why when you run `llc` with debug loggging enabled the first nodes to be legalized are things like target registers and constants, and the last node is always return. This is what allows me to use the approach I'm using, as I know that inside `WidenVecRes_CMP` the two operands have already been widened, but the CMP node itself still points to the old, unwidened operands.

If you also look into the implementation of `DAGTypeLegalizer::run`, the one of the first things it does is populate the worklist of nodes to be processed with those that have no operands, and then updates the worklist any time all arguments of some node are processed.

Reply:
Yeah, you're right. Ignore me.